File: modules/crypto/src/main/java/io/forge/jam/vrfs/BandersnatchWrapper.java
package io.forge.jam.vrfs;

import java.io.File;

/**
 * Java JNI wrapper for the Bandersnatch VRF native library.
 *
 * This class provides the JNI interface with method names that match the
 * native library's expected function signatures (Java_io_forge_jam_vrfs_BandersnatchWrapper_*).
 *
 * The native library expects this exact package and class name for JNI method resolution.
 */
public class BandersnatchWrapper {

    private static boolean libraryLoaded = false;
    private static final Object loadLock = new Object();

    /**
     * Load the native library from known paths.
     */
    public static void ensureLibraryLoaded() {
        synchronized (loadLock) {
            if (libraryLoaded) {
                return;
            }

            String libraryName = "bandersnatch_vrfs_wrapper";
            String osName = System.getProperty("os.name").toLowerCase();
            String libFileName;
            String osDirName;

            if (osName.contains("mac")) {
                libFileName = "lib" + libraryName + ".dylib";
                osDirName = "mac";
            } else if (osName.contains("linux")) {
                libFileName = "lib" + libraryName + ".so";
                osDirName = "linux";
            } else if (osName.contains("windows")) {
                libFileName = libraryName + ".dll";
                osDirName = "windows";
            } else {
                throw new RuntimeException("Unsupported operating system: " + osName);
            }

            // Get base directory from system property or use current working directory
            String baseDir = System.getProperty("jam.base.dir", System.getProperty("user.dir"));

            // Try to load from multiple possible locations
            String[] possiblePaths = {
                // From crypto module's native build output (platform-specific)
                baseDir + "/modules/crypto/native/build/" + osDirName + "/" + libFileName,
                // From crypto module's resources
                baseDir + "/modules/crypto/src/main/resources/" + libFileName,
                // From Rust target directory (direct build)
                baseDir + "/modules/crypto/native/bandersnatch-vrfs-wrapper/target/release/" + libFileName,
                // Legacy: from protocol module resources
                baseDir + "/modules/protocol/src/main/resources/" + libFileName,
                // Legacy: from protocol native-libs
                baseDir + "/modules/protocol/native-libs/" + osDirName + "/" + libFileName
            };

            boolean loaded = false;
            for (String path : possiblePaths) {
                File file = new File(path);
                if (file.exists()) {
                    try {
                        System.load(file.getAbsolutePath());
                        libraryLoaded = true;
                        loaded = true;
                        break;
                    } catch (UnsatisfiedLinkError e) {
                        // Continue to next path
                    }
                }
            }

            if (!loaded) {
                // Try loading from java.library.path as last resort
                try {
                    System.loadLibrary(libraryName);
                    libraryLoaded = true;
                } catch (UnsatisfiedLinkError e) {
                    throw new RuntimeException(
                        "Failed to load native library " + libFileName + ". Searched paths: " +
                        String.join(", ", possiblePaths), e
                    );
                }
            }
        }
    }

    /**
     * Check if the native library is loaded.
     */
    public static boolean isLibraryLoaded() {
        return libraryLoaded;
    }

    // Native method declarations - these must match the JNI function signatures
    // in the Rust library (Java_io_forge_jam_vrfs_BandersnatchWrapper_*)

    /**
     * Initialize the ring context with SRS data and ring size.
     */
    public static native byte[] initializeContext(byte[] srsData, int ringSize);

    /**
     * Get the verifier commitment (ring root) from public keys.
     */
    public static native byte[] getVerifierCommitment(int ringSize, byte[] keys);

    /**
     * Verify a ring VRF proof.
     */
    public static native byte[] verifierRingVrfVerify(
        byte[] entropy,
        long attempt,
        byte[] signature,
        byte[] commitment,
        int ringSize
    );

    /**
     * Get the VRF output from an IETF VRF signature.
     */
    public static native byte[] getIetfVrfOutput(byte[] signature);

    /**
     * Create a secret key from a seed.
     */
    public static native byte[] secretFromSeed(byte[] seed);

    /**
     * Get the public key from a secret key.
     */
    public static native byte[] publicFromSecret(byte[] secretBytes);

    /**
     * Sign using IETF VRF.
     */
    public static native byte[] ietfVrfSign(byte[] secretBytes, byte[] vrfInput, byte[] auxData);

    /**
     * Verify an IETF VRF signature.
     */
    public static native byte[] ietfVrfVerify(
        byte[] publicBytes,
        byte[] vrfInput,
        byte[] auxData,
        byte[] signature
    );

    /**
     * Get VRF output directly from secret key and input.
     */
    public static native byte[] getVrfOutput(byte[] secretBytes, byte[] vrfInput);
}

File: kotlin-archive/jam-vrfs/src/main/kotlin/io/forge/jam/vrfs/BandersnatchWrapper.kt
package io.forge.jam.vrfs

import java.io.File

class BandersnatchWrapper(ringSize: Int) {
    companion object {
        init {
            val libraryName = "bandersnatch_vrfs_wrapper"
            val osNameProperty = System.getProperty("os.name").lowercase()
            val osName = when {
                osNameProperty.contains("mac") -> "mac"
                osNameProperty.contains("linux") -> "linux"
                osNameProperty.contains("windows") -> "windows"
                else -> throw RuntimeException("Unsupported operating system: $osNameProperty")
            }

            val libFileName = when (osName) {
                "mac" -> "lib$libraryName.dylib"
                "linux" -> "lib$libraryName.so"
                "windows" -> "$libraryName.dll"
                else -> throw RuntimeException("Unsupported operating system: $osName")
            }

            val projectDir = File(System.getProperty("user.dir"))
            val parentBuildLib = projectDir.parentFile
                ?.resolve("build")
                ?.resolve("native-libs")
                ?.resolve(osName)
                ?.resolve(libFileName)

            if (parentBuildLib?.exists() == true) {
                System.load(parentBuildLib.absolutePath)
            } else {
                println("Native library not found in parent build directory")
            }
        }

        @JvmStatic
        external fun initializeContext(srsData: ByteArray, ringSize: Int): ByteArray

        @JvmStatic
        external fun getVerifierCommitment(ringSize: Int, keys: ByteArray): ByteArray?

        @JvmStatic
        external fun verifierRingVrfVerify(
            entropy: ByteArray,
            attempt: Long,
            signature: ByteArray,
            commitment: ByteArray,
            ringSize: Int
        ): ByteArray

        @JvmStatic
        external fun getIetfVrfOutput(signature: ByteArray): ByteArray

        @JvmStatic
        external fun secretFromSeed(seed: ByteArray): ByteArray

        @JvmStatic
        external fun publicFromSecret(secretBytes: ByteArray): ByteArray

        @JvmStatic
        external fun ietfVrfSign(secretBytes: ByteArray, vrfInput: ByteArray, auxData: ByteArray): ByteArray

        @JvmStatic
        external fun ietfVrfVerify(
            publicBytes: ByteArray,
            vrfInput: ByteArray,
            auxData: ByteArray,
            signature: ByteArray
        ): ByteArray

        @JvmStatic
        external fun getVrfOutput(secretBytes: ByteArray, vrfInput: ByteArray): ByteArray
    }

    init {
        val srsData = loadSrsData()
        initializeContext(srsData, ringSize)
    }

    private fun loadSrsData(): ByteArray {
        // Load from resources
        val resourcePath = "/zcash-srs-2-11-uncompressed.bin"
        return BandersnatchWrapper::class.java.getResourceAsStream(resourcePath)?.use { stream ->
            stream.readBytes()
        } ?: throw RuntimeException("SRS file not found in resources: $resourcePath")
    }


    fun verifyRingProof(
        entropy: ByteArray,
        attempt: Long,
        signature: ByteArray,
        commitment: ByteArray,
        ringSize: Int
    ): ByteArray {
        // If result is all zeros, verification failed
        try {
            val result = verifierRingVrfVerify(entropy, attempt, signature, commitment, ringSize)
            return result
        } catch (e: Exception) {
            return byteArrayOf(0)
        }
    }

    fun generateRingRoot(
        publicKeys: List<ByteArray>,
        ringSize: Int,
    ): ByteArray? {
        val srsData = loadSrsData()
        initializeContext(srsData, ringSize)
        val concatenatedKeys: ByteArray = publicKeys.flatMap { it.toList() }.toByteArray()
        return getVerifierCommitment(ringSize, concatenatedKeys)
    }
}
